docs: update mining rfc (#21)
tari-project/rfcs: src/RFC-0131_Mining.md

In hybrid mining, "independence" of algorithms is key. If the same mining hardware can be used on multiple PoW
algorithms in the hybrid mining scheme, you may as well not bother with hybrid mining because miners can simply
switch between them.

In practice, no set of algorithms is genuinely independent. The best we can do is try to choose algorithms that work best
on CPUs, GPUs, and ASICs. In truth, the distinction between GPUs and ASICs is only a matter of time. Any "GPU-friendly"
algorithm is ASIC-friendly, too; it's just a case of whether the capital outlay for fabricating them is worth it, and
this will eventually become true for any algorithm that supplies PoW for a growing market cap.

With this in mind, we should only choose one GPU/ASIC algorithm and one for CPUs. 

An excellent technical choice would be merge mining with Monero using RandomX as a CPU-only algorithm and SHA3, also known as Keccak
, for a GPU/ASIC-friendly algorithm. Using a custom configuration of such a simple and well-understood algorithm means there is 
a low likelihood of unforeseen optimizations that will give a single miner a considerable advantage. It also means that it stands a 

> good chance of being "commoditized" when ASICs are eventually manufactured. This means that SHA3 ASICs will be widely available 
> and not available from only a single supplier.

Reviewer: ```suggestion
good chance of being "commoditized" when ASICs are eventually manufactured. This would mean that SHA3 ASICs are widely available from multiple suppliers.
```

### The block distribution

> As discussed earlier, close to a 50/50 distribution is needed to reduce the chance of hash rate attacks. However,

Reviewer: ```suggestion
A 50/50 split in hash rate among algorithms minimises the chance of hash rate attacks. However,
```

sufficient buy-in is required, especially with regard to merge mining RandomX with Monero. To make it worthwhile for a

> Monero pool operator to merge mine Tari, but still, guard against hash rate attacks and to be inclusive of independent

Reviewer: ```suggestion
Monero pool operator to merge mine Tari, but still guard against hash rate attacks and to be inclusive of independent
```

> Tari supporters and enthusiasts, a 60/40 split is proposed in favor of merge mining RandomX with Monero. The

Reviewer: ```suggestion
Tari supporters and enthusiasts, a 60/40 split is proposed in favour of merge mining RandomX with Monero. The
```

take. Tari testnet has been running very successfully on Linear Weighted Moving Average (LWMA) from Bitcoin & Zcash
Clones [version 2018-11-27](https://github.com/zawy12/difficulty-algorithms/issues/3#issuecomment-442129791). This LWMA
difficulty adjustment algorithm has also been
[tested in simulations](https://github.com/tari-labs/modelling/tree/master/scenarios/multi_pow_01), and it proved to be a good choice in the multi-PoW scene as well.

### Final proposal, hybrid mining details

The final proposal is summarized below:

> - 2x mining algorithms, with an average combined target block time of 120 s, to match Monero's block interval
> - LWMA version 2018-11-27 difficulty algorithm adjustment for both with difficulty algorithim window of 90 blocks

Reviewer: ```suggestion
- Two mining algorithms, with an average combined target block time of 120 s, to match Monero's block interval.
- A log-weighted moving average difficulty adjustment algorithm using a window of 90 blocks.
```

### Tari mining hash

Tari will use a mining hash that is hashed with a 256bit Blake2b hash constructed from the Tari blockheader. 
The hash MUST be constructed with the following encoded with [consensus encoding](RFC-0121_ConsensusEncoding), in order:
- version
- previous header hash
- timestamp
- input Merkle root
- output Merkle root
- output Merkle mountain range size
- witness Merkle root
- kernel Merkle root
- kernel Merkle mountain range size
- total kernel offset
- total script offset

#### RandomX

> This will be merge mined with Monero and MUST follow all Monero consensus rules with the following for Tari. 
> The Monero coinbase transaction MUST contain the Tari mining hash in the extra field.

Reviewer: ```suggestion
Monero blocks that are merge-mining Tari MUST include the Tari mining hash in the extra field of the Monero coinbase transaction.
```

Tari also imposes the following consensus rules:
- The `seed_hash` MUST only be used for 3000 blocks, after which a block MUST be discarded if it's used again.
- The little-endian difficulty MUST be equal to or greater than the target for that block as determined by the LWMA for Tari.
- The LWMA MUST use a target time of 200 seconds.
- MUST set the header field PoW:pow_algo as 0 for a Monero block
- MUST encode the following data into the Pow:Pow_data field:
  - Monero BlockHeader,
  - RandomX VM key,
  - Monero transaction count,
  - Monero merkle root,
  - Monero coinbase merkle proof and,
  - Monero coinbase transaction

#### Sha3

The Sha3 mining hash is constructed as a Sha3-256 with the following fields:
 - Nonce
 - Tari mining hash
 - Pow data

For the Sha3 difficulty, the Sha3 mining hash is hashed with Sha3-256. 

Tari imposes the following consensus rules:
- The Big endian difficulty MUST be equal or greater than the target for that block as determined by the LWMA for Tari.

> - MUST set the header field PoW:pow_algo as 1 for a Sha block.

Reviewer: ```suggestion
- MUST set the header field PoW:pow_algo as 1 for a Sha block.
- The PoW:pow_data field is empty
```
